Silence gcc 12 and 13 warnings

wthrowe opened the pull request:
At least on my machine, there is still a runtime test failure as described in #5145.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Silence gcc out-of-bounds warning

wthrowe committed:
Silence spurious gcc warnings related to optional

wthrowe committed:
Silence gcc warning

I don't understand where this warning is coming from, as there doesn't
seem to be anything complicated going on here to hide an actual error
of confuse the analysis.

knelli2 commented on src/Options/Auto.hpp:
```diff
@@ -43,6 +43,18 @@ class Auto {
 
+  // These lines are just to work around a spurious warning.
+  Auto(Auto&&) = default;
+  Auto& operator=(Auto&&) = default;
+#if defined(__GNUC__) and not defined(__clang__) and __GNUC__ >= 12 and __GNUC__ < 13
```
Isn't this just `__GNUC__ == 12`? (if so, same for the rest of this PR)

wthrowe replied:
It is.  I often write these things as `>= first_broken and < first_optimistically_fixed` (optimistically assuming untested versions have been fixed).  In this case with only one version it's kind of silly, so I'll change it if you like.

knelli2 replied:
If you could, yeah so it's obvious it's just this one GCC version. If in the future, we find this is an issue with GCC 13 we can go back to this.

wthrowe committed:
Disable -Wdangling-reference (gcc 13)

The warning is pretty much broken.  Way more false positives than true
ones.

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=109642

wthrowe committed:
Extend gcc 12 warning suppressions to 13

wthrowe committed:
Silence (intended) gcc self-move warnings

Clang already warned.

wthrowe committed:
Fix errors caught by gcc 13 -Wpessimizing-move

wthrowe committed:
Silence -Wmaybe-uninitialized warning in YAML-cpp

wthrowe committed:
Remove pragmas for unsupported compiler versions

We require gcc 9.1 or later.

wthrowe commented:
Changed the conditional formatting, and decided while I was at this I'd do gcc 13 too (12.2.1 and 13.1.1, specifically).  Version 13 actually caught some minor bugs.

knelli2 reviewed (approved):
Looks good. Not sure what the python formatting errors are about. Is this rebased on latest develop?

kidder commented:
Ignoring python formatting issue (#5170)
